src/main/java/com/google/sps/servlets/EnterQueue.java

```diff
@@ -46,19 +46,30 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     try {
       String classCode = request.getParameter("classCode").trim();
-      Key classKey = KeyFactory.stringToKey(classCode);
-      Entity classEntity = datastore.get(classKey);
 
-      String idToken = request.getParameter("idToken");
-      FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
+      if (request.getParameter("enterTA") == null) {
+        Key classKey = KeyFactory.stringToKey(classCode);
+        Entity classEntity = datastore.get(classKey);
 
-      ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
-      updatedQueue.add(decodedToken.getUid());
-      classEntity.setProperty("studentQueue", updatedQueue);
+        String idToken = request.getParameter("idToken");
+        FirebaseToken decodedToken = authInstance.verifyIdToken(idToken);
 
-      datastore.put(classEntity);
+        ArrayList<String> updatedQueue = (ArrayList) classEntity.getProperty("studentQueue");
+        String uID = decodedToken.getUid();
+        if (!updatedQueue.contains(uID)) {
+          updatedQueue.add(uID);
```

Comment:
You need to use a `Transaction` here. These operations on the datastore must be atomic if two students try to enter at the same time.